br/pkg/lightning/restore/precheck_impl.go

```diff
@@ -0,0 +1,1164 @@
+}
+
+func NewLocalTempKVDirCheckItem(cfg *config.Config, preInfoGetter PreRestoreInfoGetter) PrecheckItem {
+	return &localTempKVDirCheckItem{
+		cfg:           cfg,
+		preInfoGetter: preInfoGetter,
+	}
+}
+
+func (ci *localTempKVDirCheckItem) GetCheckItemID() CheckItemID {
+	return CheckLocalTempKVDir
+}
+
+func (ci *localTempKVDirCheckItem) Check(ctx context.Context) (*CheckResult, error) {
+	theResult := &CheckResult{
+		Item:     ci.GetCheckItemID(),
+		Severity: Critical,
+	}
+	storageSize, err := common.GetStorageSize(ci.cfg.TikvImporter.SortedKVDir)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	localAvailable := int64(storageSize.Available)
+	estimatedDataSizeResult, err := ci.preInfoGetter.EstimateSourceDataSize(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	estimatedDataSizeWithIndex := estimatedDataSizeResult.SizeWithIndex
+
+	switch {
+	case localAvailable > estimatedDataSizeWithIndex:
+		theResult.Message = fmt.Sprintf("local disk resources are rich, estimate sorted data size %s, local available is %s",
+			units.BytesSize(float64(estimatedDataSizeWithIndex)), units.BytesSize(float64(localAvailable)))
+		theResult.Passed = true
+	case int64(ci.cfg.TikvImporter.DiskQuota) > localAvailable:
+		theResult.Message = fmt.Sprintf("local disk space may not enough to finish import, estimate sorted data size is %s,"+
+			" but local available is %s, please set `tikv-importer.disk-quota` to a smaller value than %s"+
+			" or change `mydumper.sorted-kv-dir` to another disk with enough space to finish imports",
+			units.BytesSize(float64(estimatedDataSizeWithIndex)),
+			units.BytesSize(float64(localAvailable)), units.BytesSize(float64(localAvailable)))
+		theResult.Passed = false
+		log.FromContext(ctx).Error(theResult.Message)
+	default:
+		theResult.Message = fmt.Sprintf("local disk space may not enough to finish import, "+
+			"estimate sorted data size is %s, but local available is %s,"+
+			"we will use disk-quota (size: %s) to finish imports, which may slow down import",
+			units.BytesSize(float64(estimatedDataSizeWithIndex)),
+			units.BytesSize(float64(localAvailable)), units.BytesSize(float64(ci.cfg.TikvImporter.DiskQuota)))
+		theResult.Passed = true
+		log.FromContext(ctx).Warn(theResult.Message)
+	}
+	return theResult, nil
+}
+
+type checkpointCheckItem struct {
+	cfg           *config.Config
+	preInfoGetter PreRestoreInfoGetter
+	dbMetas       []*mydump.MDDatabaseMeta
+	checkpointsDB checkpoints.DB
+}
+
+func NewCheckpointCheckItem(cfg *config.Config, preInfoGetter PreRestoreInfoGetter, dbMetas []*mydump.MDDatabaseMeta, checkpointsDB checkpoints.DB) PrecheckItem {
+	return &checkpointCheckItem{
+		cfg:           cfg,
+		preInfoGetter: preInfoGetter,
+		dbMetas:       dbMetas,
+		checkpointsDB: checkpointsDB,
+	}
+}
+
+func (ci *checkpointCheckItem) GetCheckItemID() CheckItemID {
+	return CheckCheckpoints
+}
+
+func (ci *checkpointCheckItem) Check(ctx context.Context) (*CheckResult, error) {
```

Comment:
This is actually the main loop in `(rc *Controller).DataCheck()` in `br/pkg/lightning/restore/restore.go`, L2072 - L2093, but each iteration only runs `CheckpointIsValid()`